Index: code/server/sv_client.c
===================================================================
--- code/server/sv_client.c	(revision 831)
+++ code/server/sv_client.c	(working copy)
@@ -1652,6 +1652,35 @@
 			///////////////////////////////////////////////////////////////
 			// separator for mutefix.patch and callvoteconnectprotect.patch
 			///////////////////////////////////////////////////////////////
+			if (!Q_stricmp("callvote", Cmd_Argv(0))) {
+				int callvoteRequiredConnectTime = sv_callvoteRequiredConnectTime->integer;
+				if (callvoteRequiredConnectTime > 0) {
+					if (callvoteRequiredConnectTime > 1800) { // 30 minutes.
+						callvoteRequiredConnectTime = 1800;
+					}
+					callvoteRequiredConnectTime *= 1000;
+					if (svs.time - cl->lastConnectTime < callvoteRequiredConnectTime) {
+						// Count the number of players in-game (not in spec).
+						int playersInGame = 0;
+						client_t *clTemp;
+						for (i = 0, clTemp = svs.clients; i < sv_maxclients->integer; i++, clTemp++) {
+							if (cl == clTemp || clTemp->state < CS_CONNECTED || clTemp->netchan.remoteAddress.type == NA_BOT) {
+								continue;
+							}
+							if (TEAM_SPECTATOR == atoi(Info_ValueForKey(sv.configstrings[548 + (clTemp - svs.clients)], "t"))) {
+								continue;
+							}
+							playersInGame++;
+						}
+						if (playersInGame > 0) {
+							SV_SendServerCommand(cl, "print \"You recently connected and must wait another %i seconds "
+									"before calling a vote.\n\"",
+									((callvoteRequiredConnectTime - (svs.time - cl->lastConnectTime)) / 1000) + 1);
+							return;
+						}
+					}
+				}
+			}
 			/////////////////////////////////////////////////////////////////////
 			// separator for callvoteconnectprotect.patch and cyclemaplimit.patch
 			/////////////////////////////////////////////////////////////////////
Index: code/server/server.h
===================================================================
--- code/server/server.h	(revision 831)
+++ code/server/server.h	(working copy)
@@ -314,6 +314,8 @@
 
 extern	cvar_t	*sv_userinfoFloodProtect;
 
+extern	cvar_t	*sv_callvoteRequiredConnectTime;
+
 extern	serverBan_t serverBans[SERVER_MAXBANS];
 extern	int serverBansCount;
 
Index: code/server/sv_init.c
===================================================================
--- code/server/sv_init.c	(revision 831)
+++ code/server/sv_init.c	(working copy)
@@ -691,6 +691,8 @@
 
 	sv_userinfoFloodProtect = Cvar_Get ("sv_userinfoFloodProtect", "1", CVAR_ARCHIVE | CVAR_SERVERINFO );
 
+	sv_callvoteRequiredConnectTime = Cvar_Get("sv_callvoteRequiredConnectTime", "0", CVAR_ARCHIVE);
+
 	// initialize bot cvars so they are listed and can be set before loading the botlib
 	SV_BotInitCvars();
 
Index: code/server/sv_main.c
===================================================================
--- code/server/sv_main.c	(revision 831)
+++ code/server/sv_main.c	(working copy)
@@ -61,6 +61,8 @@
 cvar_t	*sv_userinfoFloodProtect;	// whether to limit two client userinfo changes per 2000 milliseconds,
 					// default 1 (limit)
 
+cvar_t	*sv_callvoteRequiredConnectTime;
+
 serverBan_t serverBans[SERVER_MAXBANS];
 int serverBansCount = 0;
 
