Index: code/sys/con_tty.c
===================================================================
--- code/sys/con_tty.c	(revision 736)
+++ code/sys/con_tty.c	(working copy)
@@ -62,7 +62,7 @@
 
 // Note: TTY_CONSOLE_PROMPT can contain any number of characters.
 #ifdef DEDICATED
-#define TTY_CONSOLE_PROMPT "]"
+#define TTY_CONSOLE_PROMPT "ioq3ded> "
 #else
 // Right now the in-game console has the "]" prompt.
 // We choose a different prompt for the tty console for the following reasons:
