Index: code/server/sv_client.c
===================================================================
--- code/server/sv_client.c	(revision 742)
+++ code/server/sv_client.c	(working copy)
@@ -1647,6 +1647,7 @@
 			argsFromOneMaxlen = -1;
 			if (Q_stricmp("say", Cmd_Argv(0)) == 0 ||
 					Q_stricmp("say_team", Cmd_Argv(0)) == 0) {
+				sv.lastSpecChat[0] = '\0';
 				argsFromOneMaxlen = MAX_SAY_STRLEN;
 			}
 			else if (Q_stricmp("tell", Cmd_Argv(0)) == 0) {
Index: code/server/server.h
===================================================================
--- code/server/server.h	(revision 742)
+++ code/server/server.h	(working copy)
@@ -96,6 +96,8 @@
 	// separator for incognito.patch and specchatglobal.patch
 	/////////////////////////////////////////////////////////
 
+	char				lastSpecChat[512];
+
 } server_t;
 
 
@@ -300,6 +302,8 @@
 
 extern	cvar_t	*sv_userinfoFloodProtect;
 
+extern	cvar_t	*sv_specChatGlobal;
+
 extern	serverBan_t serverBans[SERVER_MAXBANS];
 extern	int serverBansCount;
 
Index: code/server/sv_init.c
===================================================================
--- code/server/sv_init.c	(revision 742)
+++ code/server/sv_init.c	(working copy)
@@ -688,6 +688,8 @@
 
 	sv_userinfoFloodProtect = Cvar_Get ("sv_userinfoFloodProtect", "1", CVAR_ARCHIVE | CVAR_SERVERINFO );
 
+	sv_specChatGlobal = Cvar_Get ("sv_specChatGlobal", "0", CVAR_ARCHIVE );
+
 	// initialize bot cvars so they are listed and can be set before loading the botlib
 	SV_BotInitCvars();
 
Index: code/server/sv_main.c
===================================================================
--- code/server/sv_main.c	(revision 742)
+++ code/server/sv_main.c	(working copy)
@@ -61,6 +61,9 @@
 cvar_t	*sv_userinfoFloodProtect;	// whether to limit two client userinfo changes per 2000 milliseconds,
 					// default 1 (limit)
 
+cvar_t	*sv_specChatGlobal;		// whether to broadcast spec chat globally
+					// default 0 don't broadcast
+
 serverBan_t serverBans[SERVER_MAXBANS];
 int serverBansCount = 0;
 
@@ -202,6 +205,15 @@
 	// separator for incognito.patch and specchatglobal.patch
 	/////////////////////////////////////////////////////////
 
+	if (sv_specChatGlobal->integer > 0 && cl != NULL &&
+			!Q_strncmp((char *) message, "chat \"^7(SPEC) ", 15)) {
+		if (!Q_strncmp((char *) message, sv.lastSpecChat, sizeof(sv.lastSpecChat) - 1)) {
+			return;
+		}
+		Q_strncpyz(sv.lastSpecChat, (char *) message, sizeof(sv.lastSpecChat));
+		cl = NULL;
+	}
+
 	if ( cl != NULL ) {
 		SV_AddServerCommand( cl, (char *)message );
 		return;
